### DISCUSSION TITLE
Error: ENOENT: no such file or directory, open '/var/task/outstatic/content/metadata.json'

### DISCUSSION THREAD

**Discussion by NorthPhoenix** [avitorio/outstatic#194] [Q&A]
Hi!

 I had Outstatic setup and stable until I recently tried to build a server-side auth with Lucia. The auth works as expected in local dev and prod, but I get this RSC error when I deploy it on Vercel:
```
[Error: ENOENT: no such file or directory, open '/var/task/outstatic/content/metadata.json'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/var/task/outstatic/content/metadata.json'
}
```

One important observation is the first instance of this error appeared after the `/blog` pathname (which is the root page for my blog) transitioned from being a static page to dynamic.

I've looked at https://github.com/avitorio/outstatic/discussions/117#discussion-5468312, and rebuilt the metadata. Error still present.

Here is my `/outstatic/content/` directory: 
![image](https://github.com/avitorio/outstatic/assets/23215805/181b0ae7-4ccd-4221-a1f0-71937810ed83)

Any help would be greatly appreciated. 🙏

**Comment by avitorio** [comment]
Hey @NorthPhoenix, 

Can you be more specific in terms of what changed in your `/blog` page?
Did you remove `generateStaticParams()`?

**Reply by NorthPhoenix** [reply]
The setup is as follows:
![image](https://github.com/avitorio/outstatic/assets/23215805/5b0a5443-c538-4f01-b146-c7166cb2dc8f)

Nothing Outstatic related changed from the previous working deployment. I only added a server Header component to `/blog/layout.tsx` that reads from cookies, which forces `/blog` route and all nested routes to be on-demand server rendered.
Here is the deployment with `Header` component commented out: https://personal-website-8kexm6wvo-northphoenix.vercel.app/blog
More detailed code is below.

```typescript
// blog/layout.tsx
const Layout = ({ children }) => {
  return (
    <Header /> // Header reads cookies for auth so this forces page to be server-rendered per request
      {children}
    <Footer />
  )
}

// blog/page.tsx
import { load } from "outstatic/server"

async function getData() {
  const blogDB = await load()
  const allPosts = await blogDB
    .find({ collection: "blog" }, [
      "title",
      "publishedAt",
      "slug",
      "coverImage",
      "description",
    ])
    .sort({ publishedAt: -1 })
    .toArray()
  return {
    allPosts,
  }
}

export default async function Index() {
  const { allPosts } = await getData()

  return (
    ...
    <ContentGrid items={allPosts} /> // just displays all posts
    ...
  )
}

// blog/[slug]/layout.tsx - nothing interesting

// blog/[slug]/page.tsx 
export default async function Post(params: { params: { slug: string } }) {
  const post = await getData(params)
  return (
    /* renders the post */
  )
}

async function getData({ params }: { params: { slug: string } }) {
  const blogDB = await load()
  const post = await blogDB
    .find<Post>({ collection: "blog", slug: params.slug }, [
      "title",
      "publishedAt",
      "description",
      "slug",
      "author",
      "content",
      "coverImage",
      "tags",
    ])
    .first()
  if (!post) {
    notFound()
  }
  const content = await markdownToHtml(post.content)
  return {
    ...post,
    content,
  }
}

export async function generateStaticParams() {
  const posts = getDocumentSlugs("blog")
  return posts.map((slug) => ({ slug }))
}


```

**Reply by avitorio** [reply]
Can you add me to you repo? I'd like to take a look at it if possible.
Does the error happen on build or when you visit the page?
The load() function should only be triggered on build time.

**Reply by NorthPhoenix** [reply]
You can definitely take a look, it's public https://github.com/NorthPhoenix/Personal-Website
Error happens on page visit. It must be the `load()` function as you mentioned, since it runs on each request.
I guess the only way to fix it is to make sure that the page is static and do an extra trip to the server to do auth.

**Reply by Pradyun-PencilHacker** [reply]
@avitorio I've been facing the same error, any suggestions as to how to debug this issue? @NorthPhoenix have you made any breakthrough in resolving this issue?

**Reply by avitorio** [reply]
The page should be static and not server-side rendered. Try searching for force-static on the next.js docs.

**Reply by FrancocDev** [reply]
I have the same problem, but in my case, I can't use the export const dynamic = 'auto' because my site loses the interactivity. The page uses parallel routes, tailwind themr based on classes and a custom Middleware to change site language based on users lang. Is there any way to implement outstatic without lose all of that?

**Reply by avitorio** [reply]
Hey @FrancocDev,
Did you find a solution?

Things can get tricky if you need to mix static and dynamic content.

**Reply by FrancocDev** [reply]
I will try it on the next days. Any way to test it in local? Because I only having problems when deploy, nothing on local, even when using "vercel dev" that's replicate the production deployment. Thanks!

**Reply by avitorio** [reply]
Hey @FrancocDev, thanks for giving this a try.

This issue seems to be happening with Vercel's edge network. Please, keep us posted.

**Reply by FrancocDev** [reply]
Hi @avitorio , I create a minimal reproduction: [FrancocDev/demo-outstatic](https://github.com/FrancocDev/demo-outstatic)
Here is the deploy: [demo-outstatic-pr14b5sbw-francocdev.vercel.app](https://demo-outstatic-pr14b5sbw-francocdev.vercel.app/)

I can't find where is the error. I tried to move the load to a RSC but I'm still getting the error. The annoying thing is that I can't debug it in local, because on local it works well. 

I don't know if outstatic is getting the wrong file path (/var/task/), because the files appears on my final build. 
Thanks in advance.

**Reply by Sardar-Abdul** [reply]
Were you guys able to solve this issue? @FrancocDev @avitorio @NorthPhoenix. Have literally tried everything but it's not working for me. Works fine in localhost, when I deploy it in vercel and click to view blog details, I am met with this exact error. Try clicking on any details of ay one of blogs here:
https://ecosmart-hotels.vercel.app/blogs

Error on vercel logs says:
Error: ENOENT: no such file or directory, open '/var/task/outstatic/content/metadata.json' (I have attached screenshot of error too).

I have tried rebuilding metadata too but it hasn't helped my cause. Can I be please helped?

![image](https://github.com/user-attachments/assets/4626184b-ee88-4f12-82df-2ab537a32848)

**Reply by NorthPhoenix** [reply]
It's been 8 months since I last tried using outstatic so my knowledge might be outdated, but my takeaway was that outstatic is built for fully static pages and cannot be used in environments where the `load` function runs not during build time (@avitorio correct me if I'm wrong).
My hope is that with time Next.js partial prerendering evolves into leaving non-dynamic parts of the page truly static and won't rebuild them on request, or outstatic evolves to handle dynamic loading. 
(It might have already happened in the last 8 months so judge that yourself)
Your blog website doesn't look like it's dynamic or needs to be dynamic, but if you need it to be dynamic I would suggest looking into different CMSs

**Reply by avitorio** [reply]
@NorthPhoenix you're correct.
The way Outstatic is setup your blog should be statically generated. One way to achieve this is to add `export const dynamic = 'force-static'` to the page.

**Reply by Sardar-Abdul** [reply]
> @NorthPhoenix you're correct. The way Outstatic is setup your blog should be statically generated. One way to achieve this is to add `export const dynamic = 'force-static'` to the page.

I added `export const dynamic = "force-static";
` at the top of my /blogs/[slug]/page.tsx file but unfortunately this doesn't solve this error for me. I get "500 - Internal Server Error." when I visit detail of any blog now. When I check my logs in vercel, I get same error (screenshot attached). 
![image](https://github.com/user-attachments/assets/c082c5a6-3d6d-42a2-83be-f6e55266a688)

What could be causing this? @avitorio.

I really appreciate you taking out time for this. Thanks.

**Reply by Sardar-Abdul** [reply]
> @NorthPhoenix you're correct. The way Outstatic is setup your blog should be statically generated. One way to achieve this is to add `export const dynamic = 'force-static'` to the page.

Thank you for this explanation.

**Reply by avitorio** [reply]
Hey @Sardar-Abdul were you ever able to get this fixed? I've seen this happen when the blog is wrapped in a context or other type of component that is client side rendered.

**Reply by hanahbaker1** [reply]
I'm getting the same issue when I deploy to production

**Reply by avitorio** [reply]
Hey @hanahbaker1, I've noticed a lot of times this issue arises because the pages are not being statically generated, have you tried adding export const dynamic = 'force-static' to the page?

**Reply by hanahbaker1** [reply]
Thanks for the suggestion @avitorio, however, that didn't quite work. On deploy the output still doesn't contain the outstatic/content folder, (not sure if it should):
![image](https://github.com/user-attachments/assets/e8377546-8afc-4f83-badd-fb5bcc99f3a6)

i can see it in the source but not the output, below is the source folder structure:
![image](https://github.com/user-attachments/assets/6f8d9c58-0231-4c03-8527-86f931dfaaa1)

the issue is definitely only for the dynamically routed content.
this is what it looks like after I've implemented the suggestion :
![image](https://github.com/user-attachments/assets/17351bf4-d3de-4932-aa1d-43a9101aba31)

**Reply by hanahbaker1** [reply]
update: it's a nextjs issue, I fixed my issue by adding this line to my next.config.mjs file:
` outputFileTracingIncludes: {
    '/articles/[slug]': ['./outstatic/**/*'],
  },`
  
  I hope this helps people with the same issue, [here](https://github.com/rehype-pretty/rehype-pretty-code/issues/98)'s the thread I found the solution from
